### ISSUE TITLE
inducing point stability issue with tensorflow optimizers 

### ISSUE THREAD

**Issue by hughsalimbeni** [GPflow/GPflow#578]
I've noticed some odd behaviour switching between optimizers:

```python
N, D = 10, 1
np.random.seed(0)
X = np.random.randn(N, D)
Y = np.random.randn(N, 1)

m_svgp = SVGP(X, Y, RBF(D), Gaussian(), X.copy())

m_svgp.set_trainable(False) 
m_svgp.q_mu.set_trainable(True)
m_svgp.q_sqrt.set_trainable(True)

ScipyOptimizer().minimize(m_svgp)  # now equivalent to gpr
m_svgp.feature.Z.set_trainable(True)  # Z is already optimal 

ScipyOptimizer().minimize(m_svgp)  # optimize some more and nothing happens, as expected
assert_allclose(m_svgp.feature.Z.read_value(), X, rtol=1e-5, atol=1e-5)  # true

GradientDescentOptimizer(0.2).minimize(m_svgp, maxiter=100)  
assert_allclose(m_svgp.feature.Z.read_value(), X, rtol=1e-5, atol=1e-5)  # false
```

I don't think this is a numerical issue, but I can't think what else can be going wrong here.

**Comment by hughsalimbeni** [comment]
Another example:

```python
N, D = 50, 1
X = np.linspace(-2, 2, N).reshape(N, 1)
Y = np.zeros((N, 1))

m_svgp = SVGP(X, Y, RBF(D), Gaussian(), X.copy())

m_svgp.set_trainable(False) 
m_svgp.q_mu.set_trainable(True)
m_svgp.q_sqrt.set_trainable(True)
ScipyOptimizer().minimize(m_svgp)  
m_svgp.feature.Z.set_trainable(True)
ScipyOptimizer().minimize(m_svgp) 
AdamOptimizer(0.1).minimize(m_svgp, maxiter=1)
plt.scatter(X, m_svgp.feature.Z.read_value())
plt.show()
```

The plot should show a straight line
[fig_adam.pdf](https://github.com/GPflow/GPflow/files/1503812/fig.pdf)

The same thing seems to happen with `whiten=False` and `use_locking=True`

With `GradientDescentOptimizer(0.1).minimize(m_svgp, maxiter=5)` it looks fine after 3 iterations, but after 5 looks like this [fig_gd.pdf](https://github.com/GPflow/GPflow/files/1503820/fig_gd.pdf)

**Comment by markvdw** [comment]
This is strange. Although in both your examples, your data is a bit pathological in the sense that I could expect not seeing particularly strong gradients from the data likelihood terms. Just to eliminate this paranoia, do you see the same if you choose the data to be lienar? E.g. `Y = np.sum(X, 1)[:, None]`?

**Comment by hughsalimbeni** [comment]
Very similar `Y=X` in the Adam example, and even more unstable with `Y = np.sum(X, 1)[:, None]` and grad descent

**Comment by markvdw** [comment]
The gradients should just be zero right? Can we calculate them directly?

**Comment by hughsalimbeni** [comment]
indeed they should, but they are not:
```python
with tf.Session() as sess:
    m_svgp.initialize()
    dZ = sess.run(tf.gradients(m_svgp.objective, m_svgp.feature.Z.unconstrained_tensor))[0]
print(dZ.flatten())
```
```
[ -8.73075723e-05   1.43944865e-05   5.65814666e-05  -8.84300926e-05
  -7.93436252e-05   1.58545579e-04   1.28330794e-04   7.16310810e-05
   4.54580132e-05  -8.76479170e-05  -1.55117446e-04  -4.20254625e-05
   4.48201747e-05   7.08789700e-05   4.64575906e-05  -1.36052424e-05
  -4.32856939e-05  -9.40420982e-06   2.27638029e-05   4.43618999e-06
  -8.77823801e-06  -2.33275043e-05  -1.00496569e-04  -9.52572597e-05
   1.13924779e-04   2.52477696e-04  -1.44898766e-05  -2.69071361e-04
  -9.30917774e-05   1.25480209e-04   1.11448204e-04   3.30424204e-05
   1.13523732e-05   7.90515329e-06  -4.74332524e-06  -1.33493282e-05
  -1.21494124e-05  -7.22056755e-06  -3.35359857e-06  -1.30961840e-06
  -4.17442203e-07  -9.00722481e-08  -1.02182132e-08  -2.69681595e-08
  -7.99549573e-08  -1.52984678e-07  -2.46929306e-07  -3.66691958e-07
  -5.16073694e-07  -6.96026645e-07]

```

**Comment by hughsalimbeni** [comment]
I've done some more looking into this and I think it is just a numerical issue. Here's a comparison of the optimized/exact and whitened/non-whitened:

```python
N, D = 100, 1
X = np.linspace(-2, 2, N).reshape(N, 1)

m_svgp_exact_w = SVGP(X, X, RBF(D), Gaussian(), X, whiten=True)
m_svgp_exact = SVGP(X, X, RBF(D), Gaussian(), X, whiten=False)
m_svgp_optimized_w = SVGP(X, X, RBF(D), Gaussian(), X, whiten=True)
m_svgp_optimized = SVGP(X, X, RBF(D), Gaussian(), X, whiten=False)

sigma2 = m_svgp_exact.likelihood.variance.read_value()
K = m_svgp_exact.kern.compute_K_symm(X)

Ky = K + np.eye(N) * sigma2
Ky_inv = np.linalg.inv(Ky)
m = K @ Ky_inv @ X
v = K - K @ Ky_inv @ K

L_inv = np.linalg.inv(np.linalg.cholesky(K + np.eye(N) * 1e-6))
mw = L_inv @ m
vw = L_inv @ v @ L_inv.T

m_svgp_exact.q_mu = m
m_svgp_exact.q_sqrt = np.linalg.cholesky(v + np.eye(N)*1e-6)[:, :, None]

m_svgp_exact_w.q_mu = mw
m_svgp_exact_w.q_sqrt = np.linalg.cholesky(vw + np.eye(N)*1e-6)[:, :, None]

for m in m_svgp_optimized_w, m_svgp_optimized:
    m.set_trainable(False)
    m.q_mu.set_trainable(True)
    m.q_sqrt.set_trainable(True)
    ScipyOptimizer().minimize(m, maxiter=10000)

fig, axs = plt.subplots(1, 2, figsize=(20, 10))

with tf.Session() as sess:
    for m, name in zip([m_svgp_exact_w, m_svgp_exact, m_svgp_optimized_w, m_svgp_optimized],
                       ['exact whitened', 'exact', 'optimized whitened', 'optimized']):
        m.initialize()

        Z = sess.run(tf.gradients(m.objective, m.feature.Z.unconstrained_tensor)[0]).flatten()
        axs[0].plot(Z, label=name)
        axs[1].plot(Z/np.std(Z), label=name)
        
axs[0].set_title('Z gradients')
axs[1].set_title('Z gradients normalized to unit standard deviation')

```

[plot.pdf](https://github.com/GPflow/GPflow/files/1504117/plot.pdf)

**Comment by hughsalimbeni** [comment]
I guess the moral is don't rely on the numerics always working perfectly. I'm still surprised that the scipy optimizer actually moves away from the correct solution, despite starting at it.

**Comment by markvdw** [comment]
Ok to summarise:
- If you initialise `SVGP` with the "exact" solution, the gradients are small.
- The whitened representation optimises to a solution that gives small gradients.
- The unwhitened case doesn't quite reach a solution with small gradients for `Z`.

Where does the scipy optimiser move away from the correct solution? I thought above the scipy solution actually stayed good?

**Comment by markvdw** [comment]
Also, parameters perhaps aren't that interesting. Hopefully the objective function value doesn't change very much. I would say that a good test for this is one where there is a large difference between the bound if you put `Z` elsewhere. Possibly a small dataset, with a reasonable large amount of noise, and data spaced reasonably far relative to the lengthscale?

You don't want to get into a situation where the inducing points are so close that it doesn't really matter where you put them in a detailed way.

**Comment by hughsalimbeni** [comment]
In addition to the stability issues above, I've now realised that Adam does bad things with small initial  gradients. This explains the odd results I was seeing (not above, but in other code that lead me to experiment with the above).

I'll put a full example here, in case anyone else might fall into this trap:

```python
import numpy as np
import tensorflow as tf
from gpflow.models import GPR, SVGP
from gpflow.training import AdamOptimizer, GradientDescentOptimizer, AdadeltaOptimizer, AdagradOptimizer, RMSPropOptimizer
from gpflow.kernels import RBF
from gpflow.likelihoods import Gaussian

Optimizers = AdamOptimizer, GradientDescentOptimizer, AdadeltaOptimizer, AdagradOptimizer, RMSPropOptimizer

N = 2
X = np.linspace(0, 1, N).reshape(N, 1)
Y = X

m_gp = GPR(X, Y, RBF(1))

model = lambda : SVGP(X, Y, RBF(1), Gaussian(), Z=X)  # fresh model each time

# find exact(ish) values
m_svgp = model()
sigma2 = m_svgp.likelihood.variance.read_value()
K = m_svgp.kern.compute_K_symm(X)
Ky = K + np.eye(N) * sigma2
Ky_inv = np.linalg.inv(Ky)
m = K @ Ky_inv @ Y
v = K - K @ Ky_inv @ K
L_inv = np.linalg.inv(np.linalg.cholesky(K + np.eye(N)*1e-6))
q_mu = L_inv @ m
q_sqrt = np.linalg.cholesky(L_inv @ v @ L_inv.T + np.eye(N)*1e-6)[:, :, None]

m_svgp.q_mu = q_mu
m_svgp.q_sqrt = q_sqrt

# check it's the same
print('gp lik {}'.format(m_gp.compute_log_likelihood()))
print('svgp lik {}'.format(m_svgp.compute_log_likelihood()))

with tf.Session() as sess:
    m_svgp.initialize()
    dZ = sess.run(tf.gradients(m_svgp.objective,
                               m_svgp.feature.Z.unconstrained_tensor)[0])
# grads should be small
print('grads Z')
print(dZ)

for Optimizer in Optimizers:
    print('\n')
    print(Optimizer)
    
    m_svgp = model()
    m_svgp.q_mu = q_mu
    m_svgp.q_sqrt = q_sqrt
    
    Optimizer(1.).minimize(m_svgp, maxiter=1)
    
    print(m_svgp.feature.Z.read_value() - X)
    
    
```

You might expect all optimizers to do not much as the gradients are quite small, but this is what happens:

```
gp lik -2.758106650855291
svgp lik -2.7581072657599366
grads Z
[[  2.05743089e-07]
 [ -3.65973141e-07]]


<class 'gpflow.training.tensorflow_optimizer.AdamOptimizer'>
[[-0.39416432]
 [ 0.53645788]]


<class 'gpflow.training.tensorflow_optimizer.GradientDescentOptimizer'>
[[ -2.05743089e-07]
 [  3.65973141e-07]]


<class 'gpflow.training.tensorflow_optimizer.AdadeltaOptimizer'>
[[ -2.05743089e-07]
 [  3.65973141e-07]]


<class 'gpflow.training.tensorflow_optimizer.AdagradOptimizer'>
[[ -6.50616775e-07]
 [  1.15730869e-06]]


<class 'gpflow.training.tensorflow_optimizer.RMSPropOptimizer'>
[[ -2.16872261e-07]
 [  3.85769568e-07]]
```

This is a problem with Adam (with small epsilon). E.g. 

```python
x_init = np.linspace(-20, 20, 10)
x = tf.Variable(x_init, dtype=tf.float64)

opt_step_adam = tf.train.AdamOptimizer(1.).minimize(-tf.exp(x))

with tf.Session() as sess:
    sess.run(tf.global_variables_initializer())
    
    sess.run(opt_step_adam)
    adam_delta = sess.run(x) - x_init

    print('gradients')
    print(np.exp(x_init))
    print('adam step size')
    print(adam_delta)
    print('expansion factor')
    print(adam_delta/np.exp(x_init))
```

```
gradients
[  2.06115362e-09   1.75512503e-07   1.49453385e-05   1.27263380e-03
   1.08368023e-01   9.22781435e+00   7.85771994e+02   6.69104951e+04
   5.69759980e+06   4.85165195e+08]
adam initial step
[  6.51747331e-05   5.51952273e-03   3.20933608e-01   9.75754023e-01
   9.99708274e-01   9.99996573e-01   9.99999960e-01   1.00000000e+00
   1.00000000e+00   1.00000000e+00]
expansion factor
[  3.16205121e+04   3.14480316e+04   2.14738266e+04   7.66720184e+02
   9.22512236e+00   1.08367652e-01   1.27263375e-03   1.49453385e-05
   1.75512503e-07   2.06115362e-09]
```

Note that when the initial gradient 1e-5 the initial step is 1e-1. This could be quite dangerous behaviour in numerically sensitive situations.

**Comment by awav** [comment]
I do believe it is Adam optimizer issue. 

```python
lr = rate * sqrt(1 - beta2^t) / (1 - beta1^t)
m = beta1 * m_prev + (1 - beta1) * g
v = beta2 * v_prev + (1 - beta2) * g^2 ```

`m` - EMA of mean gradient
`v` - EMA of squares of gradient

`v` becomes very small when we are near to optimal (in our case ==optimal), moreover `m` can change itself much more quickly, but `v` can keep its small value and therefore have dividing by almost 0:

```python
 var = var - lr * m / (sqrt(v) + e) 
```

**Comment by markvdw** [comment]
Ah yes, doesn't this behaviour cause Adam to rescale all gradients to either +1 or -1 on the first iteration?

**Comment by awav** [comment]
@markvdw, suppose that rescaling relates to different issue - high learning rate. I noticed that @hughsalimbeni used 1.0. I have never tried such large number for default beta parameters. Smaller value for learning rate is much preferable >= 0.01.

**Comment by alexggmatthews** [comment]
So are we happy this is not a problem with GPflow? If so I will close.
